use-shopping-cart/src/index.test.js

```diff
@@ -558,3 +558,97 @@ describe('stripe handling', () => {
     )
   })
 })
+
+function mockCartDetails(overrides) {
+  return {
+    [`sku_abc${counter}`]: {
+      sku: `sku_abc${counter++}`,
+      name: 'Bananas',
+      image: 'https://blah.com/banana.avif',
+      price: 400,
+      currency: 'USD',
+      value: 800,
+      quantity: 2,
+      formattedValue: '$8.00',
+      ...overrides
+    },
+    [`sku_efg${counter}`]: {
+      sku: `sku_efg${counter++}`,
+      name: 'Oranges',
+      image: 'https://blah.com/orange.avif',
+      currency: 'USD',
+      price: 250,
+      value: 1000,
+      quantity: 4,
+      formattedValue: '$10.00',
+      ...overrides
+    }
+  }
+}
+
+describe('loadCart()', () => {
+  it('should add cartDetails to cart object', async () => {
+    const wrapper = createWrapper()
+    const cart = renderHook(() => useShoppingCart(), { wrapper }).result
+    const cartDetails = mockCartDetails()
+    const product = mockProduct({ price: 200 })
+
+    act(() => {
+      cart.current.addItem(product)
+      cart.current.loadCart(cartDetails, false)
+    })
+
+    expect(cart.current.cartDetails).toEqual(cartDetails)
+    expect(cart.current.totalPrice).toEqual(1800)
+    expect(cart.current.cartCount).toEqual(6)
+  })
+
+  it('should merge two cart details items by default', async () => {
+    const wrapper = createWrapper()
+    const cart = renderHook(() => useShoppingCart(), { wrapper }).result
+
+    const cartDetails = mockCartDetails()
+    const product = mockProduct({ price: 200 })
+
+    act(() => {
+      cart.current.addItem(product)
+      cart.current.incrementItem(product.sku)
+      cart.current.loadCart(cartDetails)
+    })
+
+    const entry = cart.current.cartDetails[product.sku]
+    expect(cart.current.cartDetails).toEqual({
+      [entry.sku]: entry,
+      ...cartDetails
+    })
+    expect(cart.current.totalPrice).toEqual(2200)
+    expect(cart.current.cartCount).toEqual(8)
+  })
+})
+
+describe('filterCart', () => {
+  it('should load a filteredCart', async () => {
+    const wrapper = createWrapper()
+    const cart = renderHook(() => useShoppingCart(), { wrapper }).result
+
+    const cartDetails = mockCartDetails()
+
+    function mockFilter(item) {
+      return item.value < 1000
+    }
+
+    act(() => {
+      cart.current.loadCart(cartDetails)
+    })
+
+    let filteredCart = await cart.current.filterCart(cartDetails, mockFilter)
+
+    // doing load cart again, but filtering out one value and seeing if the end result
+    // goes back to the filteredCart
+    act(() => {
+      cart.current.loadCart(filteredCart, false)
+    })
+
+    expect(cart.current.cartDetails).toEqual(filteredCart)
+  })
+})
```

Comment:
For testing `filterCart` I would suggest just testing it standalone. The current test is only testing that `loadCart()` is working correctly and not that `filteredCart()` is.

```suggestion
describe('filterCart', () => {
	it('filter out all cart items that fail filter callback', async () => {
		const cartDetails = {
			'test-sku-123': {
				sku: 'test-sku-123',
				name: 'Bananas',
				image: 'https://blah.com/banana.avif',
				price: 200,
				currency: 'USD',
				value: 400,
				quantity: 2,
				formattedValue: '$4.00',
			},
			'test-sku-234': {
				sku: 'test-sku-234',
				name: 'Oranges',
				image: 'https://blah.com/orange.avif',
				currency: 'USD',
				price: 250,
				value: 1000,
				quantity: 4,
				formattedValue: '$10.00',
			}
		}
		
		const mockFilter = item => item.value < 500
		const filteredCart = await filterCart(cartDetails, mockFilter)

		expect(filteredCart).toEqual({
			'test-sku-123': {
				sku: 'test-sku-123',
				name: 'Bananas',
				image: 'https://blah.com/banana.avif',
				price: 200,
				currency: 'USD',
				value: 400,
				quantity: 2,
				formattedValue: '$4.00',
			},
		})
	})
})
```

This is should now utilize the utility defined `util.js`. The test should probably also be moved to `util.test.js`. I'm not certain this test works but the idea is to ensure that only items that fail the condition defined in the `filter` callback are removed.

I've also hardcoded the values for the `cartDetails` because the IDs in `mockCartDetails` are randomly generated which would make writing this test harder.

Reply:
Sorry, my editor used tabs, you can fix that by running prettier.